lib/upgrade/migrator/clients/clients.go

```diff
@@ -41,6 +41,7 @@ const (
 
 type V1ClientInterface interface {
 	Apply(d *model.KVPair) (*model.KVPair, error)
+	CmpAndSwap(k model.Key, oldValue, newValue interface{}) (*model.KVPair, error)
```

Comment:
This is a weird interface - why not just do `Update(kvpair)`? The datastore can do the compare and swap for you, so not sure why you need the old and new value here?

Reply:
We only want to update if it has the oldValue. We want this to fail if it is not the specific oldValue being provided.  I'm not sure how we achieve that with `Update(kvpair)`.

Reply:
The KVPair has a revision number which can be used for this.

Update is a CAS operation, whereas create isn't.

Reply:
Sorry I didn't realize you were talking about an already existing function (in the previous) client or I would have went and looked at that.

Do you think the Migration code is safe to assume that the Ready flag is present? If not then I'll have to add back in the Create function to the migration clientV1 interface.

Reply:
eh, didn't realize all of that code has been ripped out now... 

> Do you think the Migration code is safe to assume that the Ready flag is present?

Do you? I sorta think it might be, but I'm not close enough to the migration to know for sure. Maybe Rob would know.

Alternatively, can't we store this state as an annotation on e.g. the k8s kube-system namespace or something and clear it when we're done? Just another thought (feel free to ignore if it's not any easier)

Reply:
Yeah I think it is safe to expect that it is there. It has been created by EnsureInitialized since v2.4.1 (and possible earlier).